atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/store/AtlasDbTransactionStore.java

```diff
@@ -0,0 +1,181 @@
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+public final class AtlasDbTransactionStore implements TransactionStore {
+
+    private static final SafeLogger log = SafeLoggerFactory.get(AtlasDbTransactionStore.class);
+    private final TransactionManager transactionManager;
+    private final TableReference tableReference;
+    private final TableReference indexTableReference;
+
+    private AtlasDbTransactionStore(
+            TransactionManager transactionManager, TableReference tableReference, TableReference indexTableReference) {
+        this.transactionManager = transactionManager;
+        this.tableReference = tableReference;
+        this.indexTableReference = indexTableReference;
+    }
+
+    @Override
+    public Optional<Integer> get(WorkloadCell cell) {
+        Cell atlasCell = toAtlasCell(cell);
+        return transactionManager.runTaskWithRetry(task -> {
+            Map<Cell, byte[]> values = task.get(tableReference, Set.of(atlasCell));
+            return Optional.ofNullable(values.get(atlasCell)).map(Ints::fromByteArray);
+        });
+    }
+
+    @Override
+    public Optional<WitnessedTransaction> readWrite(List<TransactionAction> actions) {
+        AtomicReference<Long> startTimestampReference = new AtomicReference<>();
+        AtomicReference<List<WitnessedTransactionAction>> witnessedActionsReference = new AtomicReference<>();
+        try {
+            transactionManager.runTaskWithRetry(txn -> {
+                AtlasDbTransactionActionVisitor visitor = new AtlasDbTransactionActionVisitor(txn);
+                startTimestampReference.set(txn.getTimestamp());
+                witnessedActionsReference.set(actions.stream()
+                        .sequential()
+                        .map(action -> action.accept(visitor))
+                        .collect(Collectors.toList()));
+                return null;
+            });
+
+            TransactionStatus status =
+                    transactionManager.getTransactionService().getV2(startTimestampReference.get());
+
+            Optional<Long> commitTimestamp = TransactionStatuses.getCommitTimestamp(status);
+
+            return Optional.of(ImmutableWitnessedTransaction.builder()
+                    .startTimestamp(startTimestampReference.get())
+                    .commitTimestamp(commitTimestamp)
+                    .actions(witnessedActionsReference.get())
+                    .build());
+        } catch (Exception e) {
+            log.info("Failed to record transaction due to an exception", e);
+            return Optional.empty();
+        }
+    }
+
+    private class AtlasDbTransactionActionVisitor implements TransactionActionVisitor<WitnessedTransactionAction> {
```

Comment:
Opt'd for non-static so we've a reference to the internal methods/classes. Maybe having the AtlasDbStore itself implement `TransactionActionVisitor<WitnessedTransactionAction>` would be cleaner, but I personally find it a bit yuck.

Reply:
Agree with you on keeping the processing function separate!